### PR TITLE
chore: Simplify CometShuffleMemoryAllocator to use Spark unified memory allocator

### LINKED ISSUE #886: SparkOutOfMemoryError happens when running CometColumnarExchange
### Describe the bug

We easily run into this problem when running queries with `spark.comet.exec.shuffle.mode=jvm`:

```
org.apache.spark.memory.SparkOutOfMemoryError: Unable to acquire 67108848 bytes of memory, got 65700208 bytes. Available: 65700208

org.apache.spark.memory.SparkOutOfMemoryError: Unable to acquire 67108848 bytes of memory, got 65700208 bytes. Available: 65700208
	at org.apache.spark.shuffle.comet.CometShuffleMemoryAllocator.allocate(CometShuffleMemoryAllocator.java:132)
	at org.apache.spark.shuffle.comet.CometShuffleMemoryAllocator.allocatePage(CometShuffleMemoryAllocator.java:119)
	at org.apache.spark.sql.comet.execution.shuffle.SpillWriter.initialCurrentPage(SpillWriter.java:158)
	at org.apache.spark.shuffle.sort.CometShuffleExternalSorter.insertRecord(CometShuffleExternalSorter.java:368)
	at org.apache.spark.sql.comet.execution.shuffle.CometUnsafeShuffleWriter.insertRecordIntoSorter(CometUnsafeShuffleWriter.java:278)
	at org.apache.spark.sql.comet.execution.shuffle.CometUnsafeShuffleWriter.write(CometUnsafeShuffleWriter.java:206)
	at org.apache.spark.shuffle.ShuffleWriteProcessor.write(ShuffleWriteProcessor.scala:59)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:101)
	at org.apache.spark.scheduler.ShuffleMapTask.runTask(ShuffleMapTask.scala:53)
	at org.apache.spark.TaskContext.runTaskWithListeners(TaskContext.scala:161)
	at org.apache.spark.scheduler.Task.run(Task.scala:139)
	at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:554)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1529)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:557)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source)
	at java.base/java.lang.Thread.run(Unknown Source)

```

We've observed this problem not only on our own workloads but also on TPC-H benchmarks. The above-mentioned exception happens when running [TPC-H query 5](https://github.com/apache/datafusion-benchmarks/blob/main/tpch/queries/q5.sql) on parquet files with scale factor = 1000.

We've tried to disable the comet shuffle manager and use Spark's own shuffle exchange, all TPC-H queries could finish successfully.

### Steps to reproduce

Running TPC-H query 5 on a Spark cluster. The detailed environment and spark configurations are listed in **Additional context**.

### Expected behavior

All TPC-H queries should finish successfully.

### Additional context

The problem was produced on a self-deployed K8S Spark cluster on AWS.

* Driver/executor instance type: r7i.2xlarge (8 vCPUs, 64GB memory)
* Executor pod resource limit: 6 vCPUs, 48GB memory. We reserved some resources for some reason
* Number of executor instances: 48
* Spark version: 3.4.0
* Java version: 17
* Comet version: commit https://github.com/apache/datafusion-comet/commit/9205f0d1913933f2cc8767c02a7728a4e318dd49

Here are relevant spark configurations:

```
spark.executor.cores 6
spark.executor.memory 30719m
# Reserve native memory for comet, python and other stuff
spark.executor.memoryOverheadFactor 0.6
# Each executor core gets 1.2 GB memory for comet, all 6 executors will use 7.2GB memory.
# I know this is too small for comet, but it should not prevent the query from finishing
spark.comet.memory.overhead.factor 0.04

spark.sql.extensions org.apache.comet.CometSparkSessionExtensions
spark.comet.enabled true
spark.comet.exec.enabled true
spark.comet.exec.all.enabled true
spark.comet.exec.shuffle.enabled true
spark.comet.exec.shuffle.mode jvm
spark.shuffle.manager org.apache.spark.sql.comet.execution.shuffle.CometShuffleManager
```

### LINKED ISSUE #1064: CometShuffleMemoryAllocator should use Spark unified  memory manager
### What is the problem the feature request solves?

Currently Comet JVM shuffle's CometShuffleMemoryAllocator uses a separate unsafe memory allocator and track off-heap memory allocation internally. The memory size is configured by a Comet config.

A better approach should be to use Spark unified  memory manager for CometShuffleMemoryAllocator. So we can unify this off-heap memory with Spark off-heap memory config, i.e., we don't need to separately configure Comet JVM shuffle memory.

### Describe the potential solution

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by viirya**
## Which issue does this PR close?

Closes #1064 
Closes #886

### CODE REVIEW

**Comment by andygrove** [comment]
test failure:

```
 [info] - Spark vectorized reader - with partition data column - select a single complex field from a map entry and its parent map entry *** FAILED *** (653 milliseconds)
[info]   org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 215.0 failed 1 times, most recent failure: Lost task 0.0 in stage 215.0 (TID 370) (4bf8ef4698e6 executor driver): java.lang.IllegalArgumentException: CometShuffleMemoryAllocator should be used with off-heap memory mode, but got ON_HEAP
[info] 	at org.apache.spark.shuffle.comet.CometShuffleMemoryAllocator.getInstance(CometShuffleMemoryAllocator.java:44)
[info] 	at org.apache.spark.sql.comet.execution.shuffle.CometDiskBlockWriter.<init>(CometDiskBlockWriter.java:139)
[info] 	at org.apache.spark.sql.comet.execution.shuffle.CometBypassMergeSortShuffleWriter.write(CometBypassMergeSortShuffleWriter.java:181)
```

I think we need to specify `spark.memory.offHeap.enabled=true` when running Spark tests? I need to do the same in https://github.com/apache/datafusion-comet/pulls

For this PR we should also fall back to Spark for shuffle if `spark.memory.offHeap.enabled=false`?

**Comment by viirya** [comment]
Basically Spark tests are running with on-heap config, except for tests that particularly for off-heap test.

I'm not sure if enabling off-heap for all Spark tests can pass them all. If it works, let's do it.

If not, I plan to keep and rename current CometShuffleMemoryAllocator to a test-only class CometTestShuffleMemoryAllocator. Once it runs Spark tests, Comet can use CometTestShuffleMemoryAllocator to run Spark tests.

**Comment by viirya** [comment]
@andygrove All Spark tests are passed now.

**Comment by andygrove** [comment]
I tried testing with TPC-H but see a memory issue:

```
│ 24/11/08 02:31:44 INFO core/src/lib.rs: Comet native library version 0.4.0 initialized                                                                                                                        │
│ #                                                                                                                                                                                                             │
│ # A fatal error has been detected by the Java Runtime Environment:                                                                                                                                            │
│ #                                                                                                                                                                                                             │
│ #  SIGSEGV (0xb) at pc=0x00007399e4661564, pid=11, tid=132                                                                                                                                                    │
│ #                                                                                                                                                                                                             │
│ # JRE version: OpenJDK Runtime Environment Temurin-11.0.24+8 (11.0.24+8) (build 11.0.24+8)                                                                                                                    │
│ # Java VM: OpenJDK 64-Bit Server VM Temurin-11.0.24+8 (11.0.24+8, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)                                                                           │
│ # Problematic frame:                                                                                                                                                                                          │
│ corrupted double-linked list             
```

**Comment by andygrove** [comment]
One other issue. I tested with `spark.memory.offHeap.enabled=false` and the shuffle did not fall back to Spark but failed at runtime.

**Comment by viirya** [comment]
> I tried testing with TPC-H but see a memory issue:

I will test it locally too.

**Comment by viirya** [comment]
> One other issue. I tested with `spark.memory.offHeap.enabled=false` and the shuffle did not fall back to Spark but failed at runtime.

Yes. If using on-heap config, CometShuffleMemoryAllocator will throw runtime error.

The test only CometTestShuffleMemoryAllocator is only used in Spark/Comet tests.

**Comment by viirya** [comment]
> I tried testing with TPC-H but see a memory issue:

Hmm, I just ran TPC-H with this PR on Spark 3.4 using datafusion-comet script without any error.

**Comment by andygrove** [comment]
> Yes. If using on-heap config, CometShuffleMemoryAllocator will throw runtime error, i.e., you need to use off-heap config in Spark.

Right, so if the user is using on-heap, we should not use Comet shuffle and should fall back to Spark. We probably just need to update `isCometShuffleEnabled` to check if off-heap is being used.

**Comment by viirya** [comment]
> > Yes. If using on-heap config, CometShuffleMemoryAllocator will throw runtime error, i.e., you need to use off-heap config in Spark.
> 
> Right, so if the user is using on-heap, we should not use Comet shuffle and should fall back to Spark. We probably just need to update `isCometShuffleEnabled` to check if off-heap is being used.

Oh, I see. That sounds good. I will update it.

**Comment by viirya** [comment]
I don't set `spark.comet.memory.overhead.factor`. Do you need it?

**Comment by andygrove** [comment]
Actually, this PR won't close #886 because this is still using a singleton, so let's ignore that for now.

This PR LGTM and I will approve after some more testing.

**Comment by viirya** [comment]
It fallbacks to Spark shuffle now if off-heap is not enabled.

**Review by andygrove — APPROVED** [review]
Thanks @viirya

**Comment by viirya** [comment]
> Actually, this PR won't close #886 because this is still using a singleton, so let's ignore that for now.

As now the allocator uses all available memory on the executor (we don't specify memory size on the allocator), it should not be an issue for #886 now. @andygrove Do you want to re-check if #886 can be fixed by this PR too? Thanks.

**Comment by andygrove** [comment]
Can we make `COMET_COLUMNAR_SHUFFLE_MEMORY_SIZE` and `COMET_COLUMNAR_SHUFFLE_MEMORY_FACTOR` internal configs now, since they are only used in tests now?

**Comment by viirya** [comment]
> Can we make `COMET_COLUMNAR_SHUFFLE_MEMORY_SIZE` and `COMET_COLUMNAR_SHUFFLE_MEMORY_FACTOR` internal configs now, since they are only used in tests now?

Yes. They should be internal configs now.

**Comment by andygrove** [comment]
> As now the allocator uses all available memory on the executor (we don't specify memory size on the allocator), it should not be an issue for #886 now. @andygrove Do you want to re-check if #886 can be fixed by this PR too? Thanks.

I will test this again today.

**Comment by andygrove** [comment]
I'm running into SIGSEGV issues again.

```
│ # A fatal error has been detected by the Java Runtime Environment:                                                                                                                                      │
│ #                                                                                                                                                                                                       │
│ #  SIGSEGV (0xb) at pc=0x000072e2c93b6bc8, pid=11, tid=127                                                                                                                                              │
│ #                                                                                                                                                                                                       │
│ # JRE version: OpenJDK Runtime Environment Temurin-11.0.24+8 (11.0.24+8) (build 11.0.24+8)                                                                                                              │
│ # Java VM: OpenJDK 64-Bit Server VM Temurin-11.0.24+8 (11.0.24+8, mixed mode, sharing, tiered, compressed oops, g1 gc, linux-amd64)                                                                     │
│ # Problematic frame:                                                                                                                                                                                    │
│ # C  [libcomet-14210005976568904946.so+0x736bc8]  comet::execution::shuffle::row::append_columns::h9b53b563e484a30e+0x1318                                                                              │
│ #      
```

I will try running the same benchmark on main.

**Comment by andygrove** [comment]
I increased the off-heap pool size, and now I can run TPC-H q5 @ sf=1TB on the `main` branch, but get SIGSEGV with this PR.

**Comment by viirya** [comment]
Let me see if I can reproduce it.

**Comment by viirya** [comment]
Ah, I figured out what was wrong there. I updated this with the change.

Please also run the benchmarks to verify it. Thanks. @andygrove

File: spark/src/main/java/org/apache/spark/shuffle/comet/CometShuffleMemoryAllocator.java [hunk 1]
```diff
@@ -20,179 +20,86 @@
...
+      if (taskMemoryManager.getTungstenMemoryMode() != MemoryMode.OFF_HEAP) {
+        throw new IllegalArgumentException(
+            "CometShuffleMemoryAllocator should be used with off-heap "
+                + "memory mode, but got "
+                + taskMemoryManager.getTungstenMemoryMode());
+      }
+
+      // CometShuffleMemoryAllocator stores pages in TaskMemoryManager which is not singleton,
+      // but one instance per task. So we need to create a new instance for each task.
+      return new CometShuffleMemoryAllocator(taskMemoryManager, pageSize);
```
Comment (andygrove) [inline spark/src/main/java/org/apache/spark/shuffle/comet/CometShuffleMemoryAllocator.java:72]: I think this should also address my concerns about https://github.com/apache/datafusion-comet/issues/886. I am testing now.

File: spark/src/main/java/org/apache/spark/shuffle/comet/CometShuffleMemoryAllocator.java [hunk 2]
```diff
@@ -20,179 +20,86 @@
...
-  private static CometShuffleMemoryAllocator INSTANCE;
-
-  public static synchronized CometShuffleMemoryAllocator getInstance(
+  /**
+   * Returns the singleton instance of `CometShuffleMemoryAllocator`. This method should be used
+   * instead of the constructor to ensure that only one instance of `CometShuffleMemoryAllocator` is
+   * created. For Spark tests, this returns `CometTestShuffleMemoryAllocator` which is a test-only
+   * allocator that should not be used in production.
+   */
+  public static synchronized CometShuffleMemoryAllocatorTrait getInstance(
```
Comment (andygrove) [inline spark/src/main/java/org/apache/spark/shuffle/comet/CometShuffleMemoryAllocator.java:48]: Just a nit, but we could stop making the method synchronized and add a synchronized block around the `INSTANCE` creation when in test mode?

Comment (viirya) [reply spark/src/main/java/org/apache/spark/shuffle/comet/CometShuffleMemoryAllocator.java:48]: Okay, sounds good. I will update this.

**Review by andygrove — APPROVED** [review]
I no longer see memory errors with the recent changes.

**Comment by viirya** [comment]
Cool. Thanks @andygrove for verifying it.

**Pull request merged**